File: day07/sizer.rb
require "pry"
module Sizer
  @dirs = {}
  @current_directory = nil

  class Directory
    attr_accessor :name, :parent, :files, :children

    def initialize(attrs = {})
      self.name = attrs[:name]
      self.parent = attrs[:parent]
      self.files = attrs.fetch(:files, [])
      self.children = attrs.fetch(:children, [])
    end

    def size_sum
      files.sum(&:size) + children.sum(&:size_sum)
    end
  end

  class File
    attr_accessor :name, :parent, :size

    def initialize(attrs = {})
      self.name = attrs[:name]
      self.parent = attrs[:parent]
      self.size = attrs[:size]
    end
  end

  def self.run(filepath = 'test_input.txt')
    populate_dirs(filepath)
    puts size_sum_of_dirs_with_size_at_most(100000)
  end

  def self.size_sum_of_dirs_with_size_at_most(max_size)
    @dirs.values.map { |dir| dir.size_sum }.reject { |size_sum| size_sum > max_size}.sum
  end

  def self.populate_dirs(filepath = 'test_input.txt')
    ::File.readlines(filepath).each do |line|
      puts line
      formatted_line = line.split
      case formatted_line[0]
      when "$" # command
        case formatted_line[1]
        when "cd"
          process_cd(formatted_line[2])
        when "ls"
          next
        end
      when "dir" # listing a directory
        find_or_create_directory(formatted_line[1])
      when /\d+/
        find_or_create_file(formatted_line[0], formatted_line[1])
      end
    end
  end

  def self.process_cd(dir_name)
    if dir_name == ".."
      @current_directory = @current_directory.parent || @current_directory # in case they're already at root
    else
      dir = find_or_create_directory(dir_name)
      @current_directory = dir
    end
  end

  def self.find_or_create_directory(dir_name)
    dir = @dirs[dir_name] || Directory.new(name: dir_name, parent: @current_directory)
    @dirs[dir_name] ||= dir
    @current_directory.children << dir if @current_directory
    dir
  end

  def self.find_or_create_file(file_size, file_name)
    return if @current_directory.files.find { |f| f.name == file_name}
    file = File.new(name: file_name, size: file_size.to_i, parent: @current_directory)
    @current_directory.files << file
  end
end

Sizer.run("input.txt")

File: day07/sizer.ex
defmodule Sizer do
  defmodule Directory do
    defstruct [:parent, :children, :name, :files]
  end

  defmodule AFile do
    defstruct [:parent, :name, :size]
  end

  # Example:
  # dirs = {
  #   "/" => 48381165
  #   "a" => 94853
  #   "e" => 584
  #   "d" => 24933642
  # }

  def run(filepath \\ "test_input.txt") do
    dirs = populate_dirs_for_file(filepath)
    IO.inspect dirs
  end

  def populate_dirs_for_file(filepath) do
    File.read!(filepath)
    |> String.split("\n")
    |> populate_dirs()
  end

  ######## populate_dirs
  defp populate_dirs([line | remaining_lines]) do # Before there is a current_directory
    {current_directory, dirs} = String.split(line) |> process_line_parts(%{}, nil)

    populate_dirs(remaining_lines, dirs, current_directory)
  end
  defp populate_dirs([line | remaining_lines], dirs, current_directory) do
    IO.puts("LINE: #{line}")
    {current_directory, dirs} = String.split(line) |> process_line_parts(dirs, current_directory)

    populate_dirs(remaining_lines, dirs, current_directory)
  end

  defp populate_dirs([], dirs, _current_directory), do: dirs

  ######## process_line_parts
  defp process_line_parts([], dirs, current_directory), do: {current_directory, dirs}
  defp process_line_parts(["$", "cd", ".."], dirs, current_directory) do
    {current_directory.parent, dirs}
  end

  defp process_line_parts(["$", "cd", dir_name], dirs, current_directory) do
    find_or_create_dir(dir_name, dirs, current_directory)
  end

  defp process_line_parts(["$", "ls"], dirs, current_directory), do: {current_directory, dirs}

  defp process_line_parts(["dir", dir_name], dirs, current_directory) do
    {_dir, dirs} = find_or_create_dir(dir_name, dirs, current_directory)
    {current_directory, dirs}
  end

  defp process_line_parts([file_size, file_name], dirs, current_directory) do
    file = struct(AFile, parent: current_directory, name: file_name, size: file_size)

    # if file_size == "8504156" do
    # require IEx; IEx.pry
    # end

    case Enum.member?(current_directory.files, file) do
      true -> {current_directory, dirs}
      false ->
        # require IEx; IEx.pry
        {Map.replace(current_directory, :files, [file | current_directory.files]), dirs}
    end
  end

  defp find_or_create_dir(dir_name, %{}, nil) do
    dir = struct(Directory, parent: nil, name: dir_name, children: [], files: [])

    {dir, %{dir_name => dir}}
  end
  defp find_or_create_dir(dir_name, dirs, current_directory) do
    case dirs[dir_name] do
      nil ->
        dir = struct(Directory, parent: current_directory, name: dir_name, children: [], files: [])

        current_directory = Map.put(current_directory, :children, [dir | current_directory.children])

        {dir, Map.put(dirs, dir_name, dir)}
      _dir ->
        {dirs[dir_name], dirs}
    end
  end
end

Sizer.run("test_input.txt")
